src/serial.rs

```diff
@@ -551,6 +581,37 @@ macro_rules! usart {
                     self.usart.cr2.reset();
                     self.usart.cr3.reset();
 
+                    let fifo_threshold_bits = match(config.rxfifothreshold){
+                        FifoThreshold::OneEighth => 0,
+                        FifoThreshold::OneQuarter => 1,
+                        FifoThreshold::OneHalf => 2,
+                        FifoThreshold::TreeQuarter => 3,
+                        FifoThreshold::SevenEighth => 4,
+                        FifoThreshold::Full => 5,
+                    };
+
+                    unsafe {
+                        self.usart.cr3.write(|w| {w.rxftcfg().bits(fifo_threshold_bits)});
+                    }
+
+
+
+                    let fifo_threshold_bits = match(config.txfifothreshold){
+                        FifoThreshold::OneEighth => 0,
+                        FifoThreshold::OneQuarter => 1,
+                        FifoThreshold::OneHalf => 2,
+                        FifoThreshold::TreeQuarter => 3,
+                        FifoThreshold::SevenEighth => 4,
+                        FifoThreshold::Full => 5,
+                    };
+
+
+                    unsafe{
+                        self.usart.cr3.write(|w| {w.txftcfg().bits(fifo_threshold_bits) });
```

Comment:
You should use the `modify` method instead of `write` here. Otherwise the rxftcfg bit will be overwritten!

To avoid this in general, I would actually suggest using `modify` instead of `write` in both cases.

```suggestion
                        self.usart.cr3.modify(|_, w| {w.txftcfg().bits(fifo_threshold_bits)});
```

Reply:
Oh thanks. Careless mistake
One question more, I begin Rust, could you tell me why on some registers it is mandatory to put code in safe section meanwhile others register can be modify outside unsafe section?

Reply:
It's a good question, the answer is in the svd2rust documentation [on this page](https://docs.rs/svd2rust/latest/svd2rust/#write)

> // NOTE ^ unsafe because you could be writing a reserved bit pattern into
// the register. In this case, the SVD doesn't provide enough information to
// check whether that's the case.

So it depends if someone has contributed a definition of the valid (not reserved) bit patterns to the [stm32-rs project](https://github.com/stm32-rs/stm32-rs). If there is a definition then it will be safe, otherwise unsafe is required.

Reply:
Ok thanks you for your complete answer